backend/src/main/java/de/bund/digitalservice/ris/norms/domain/entity/EId.java

```
@@ -50,4 +66,59 @@ public static Optional<EId> fromNode(Node node) {
   public static EId fromMandatoryNode(Node node) {
     return new EId(NodeParser.getValueFromMandatoryNodeFromExpression("./@eId", node));
   }
+
+  /**
+   * Creates the expected EId for the given node given that the parent nodes eId is correct.
+   *
+   * <p>See LDML.de 1.6 Section 7.1.1
+   *
+   * @param node the node for which the eId should be calculated
+   * @return the expected EId for that node
+   */
+  public static Optional<EId> forNode(Node node) {
+    final Optional<EId> parentEId = EId.fromNode(node.getParentNode());
+
+    final Optional<EIdPartType> eIdPartType = EIdPartType.forAknElement(node);
+
+    if (eIdPartType.isEmpty()) {
+      return Optional.empty();
+    }
+
+    var newEIdPart =
+        new EIdPart(eIdPartType.get().getName(), findEIdPosition(node, eIdPartType.get()));
+
+    return parentEId
+        .map(eId -> eId.addPart(newEIdPart))
+        .or(() -> Optional.of(new EId(newEIdPart.value())));
+  }
+
+  private static String findEIdPosition(Node node, EIdPartType eIdPartType) {
+    if (node.getNodeName().equals("akn:li")
+        && node.getParentNode().getNodeName().equals("akn:ol")) {
+      return node.getAttributes().getNamedItem("value").getNodeValue().replace(".", "");
+    }
+
+    return NodeParser.getValueFromExpression("./num/marker", node)
+        .or(() -> NodeParser.getValueFromExpression("./num/marker/@name", node))
+        .orElseGet(
+            () -> {
+              var position = 1;
+              var previousSibling = node.getPreviousSibling();
+              while (previousSibling != null) {
+                var eId = EId.fromNode(previousSibling);
+
+                if (eId.isPresent()) {
```

Should we throw an exception here if we have a previous sibling but that sibling does not have an eId?

---

i dont think so as it could be fine that a previous sibling has no eid if it is an element that does not need an eid

---

You are right. And also the way the code is written, if that previous sibling didn't have an eId if would have gotten one in the loop run before, right? You can resolve the conversation

---

yes if it is used in the EIdConsistencyGuard it would have already gotten one